### PR TITLE
Implement feature that allows to export tabular data suited to a terminal. This solves issue #340.

### PR DESCRIPTION
**Pull request by xdanielsb**
Hi I implemented a wrapper that helps to export tabular data into a Command Line Interface. All test with different versions of  formats were performed. I appreciate any comment or suggestion. If there is any additional  change or comment, please let me know.

### CODE REVIEW

File: src/tablib/formats/_cli.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+class CLIFormat:
+    """ Class responsible to export to CLI Format """
+    title = 'cli'
+    DEFAULT_FMT = 'plain'
+
+    @classmethod
+    def export_set(cls, dataset, **kwargs):
+        """Returns CLI representation of a Dataset."""
+        if( dataset.headers ):
```
Comment (hugovk) [inline src/tablib/formats/_cli.py]:
```suggestion
        if (dataset.headers):
```

File: tests/test_tablib.py [hunk 2]
```diff
@@ -1288,3 +1288,29 @@ def test_rst_formatter_doctests(self):
...
         results = doctest.testmod(tablib.formats._rst)
         self.assertEqual(results.failed, 0)
+
+
+class CliTests(BaseTestCase):
+    def test_cli_export(self):
+        self.assertEqual('a  b  c', tablib.Dataset(['a', 'b', 'c']).cli)
+
+    def test_cli_export_github(self):
+        self.assertEqual('|---|---|---|\n| a | b | c |', tablib.Dataset(['a','b','c']).export('cli',tablefmt='github'))
```
Comment (hugovk) [inline tests/test_tablib.py]: Here and below:
```suggestion
        self.assertEqual('|---|---|---|\n| a | b | c |', tablib.Dataset(['a','b','c']).export('cli', tablefmt='github'))
```

File: setup.py [hunk 3]
```diff
@@ -33,12 +33,13 @@
     ],
     python_requires='>=3.5',
     extras_require={
-        'all': ['markuppy', 'odfpy', 'openpyxl>=2.4.0', 'pandas', 'pyyaml', 'xlrd', 'xlwt'],
+        'all': ['markuppy', 'odfpy', 'openpyxl>=2.4.0', 'pandas', 'pyyaml', 'xlrd', 'xlwt', 'tabulate'],
```
Comment (hugovk) [inline setup.py]: Please put these in alphabetical order, plus `'cli'` below.

File: src/tablib/formats/__init__.py [hunk 4]
```diff
@@ -36,6 +36,11 @@
         "The 'yaml' format is not available. You may want to install the pyyaml "
         "package (or `pip install tablib[yaml]`)."
     ),
+    'cli': (
+        "The 'cli' format is not available. You may want to install the tabulate "
+        "package (or `pip install tablib[cli]`)."
+    ),
+
```
Comment (claudep) [inline src/tablib/formats/__init__.py]: Blank line unneeded.

File: src/tablib/formats/__init__.py [hunk 5]
```diff
@@ -124,6 +129,7 @@ def register_builtins(self):
         if find_spec('pandas'):
             self.register('df', 'tablib.formats._df.DataFrameFormat')
         self.register('rst', 'tablib.formats._rst.ReSTFormat')
+        self.register('cli', 'tablib.formats._cli.CLIFormat')
```
Comment (claudep) [inline src/tablib/formats/__init__.py]: This line should also be conditionally registered, depending on installed packages.

File: src/tablib/formats/_cli.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
+"""Tablib - Command-line Interface table export support.
+
+   Generates a representation for CLI from the dataset.
+   Wrapper for tabulate library
+"""
+try:
+    from tabulate import tabulate as Tabulate
+except ImportError:
+    Tabulate = None
```
Comment (claudep) [inline src/tablib/formats/_cli.py]: No need for that conditional import. If tabulate is not available, the import error should bubble up.

File: src/tablib/formats/_cli.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+class CLIFormat:
+    """ Class responsible to export to CLI Format """
+    title = 'cli'
+    DEFAULT_FMT = 'plain'
+
+    @classmethod
+    def export_set(cls, dataset, **kwargs):
+        """Returns CLI representation of a Dataset."""
+        if (dataset.headers):
```
Comment (claudep) [inline src/tablib/formats/_cli.py]: The parens are not needed.

File: src/tablib/formats/_cli.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+            kwargs.setdefault('headers', dataset.headers)
+        kwargs.setdefault('tablefmt', cls.DEFAULT_FMT)
+        return Tabulate( dataset, **kwargs)
```
Comment (claudep) [inline src/tablib/formats/_cli.py]: No space after parens (same in the next method).

File: tests/test_tablib.py [hunk 9]
```diff
@@ -1288,3 +1288,29 @@ def test_rst_formatter_doctests(self):
...
+        self.assertEqual('+---+---+---+\n| a | b | c |\n+---+---+---+', tablib.Dataset(['a','b','c']).export('cli', tablefmt='grid'))
+
+    def test_cli_export_presto(self):
+        self.assertEqual(' a | b | c', tablib.Dataset(['a','b','c']).export('cli', tablefmt='presto'))
+
+    def test_cli_export_moinmoin(self):
+        self.assertEqual('||  a  ||  b  ||  c  ||', tablib.Dataset(['a','b','c']).export('cli', tablefmt='moinmoin'))
+
+    def test_cli_export_youtrack(self):
+        self.assertEqual('|  a  |  b  |  c  |', tablib.Dataset(['a','b','c']).export('cli', tablefmt='youtrack'))
```
Comment (claudep) [inline tests/test_tablib.py]: I dont think we need to test all those formats. We don't aim to test tabulate, but tablib, so one or two formats should be enough. Apparently, `export_book` is not tested.

File: docs/formats.rst [hunk 10]
```diff
@@ -7,6 +7,22 @@ Formats
 Tablib supports a wide variety of different tabular formats, both for input and
 output. Moreover, you can :ref:`register your own formats <newformats>`.
 
+cli
+====
```
Comment (hugovk) [inline docs/formats.rst]:
```suggestion
===
```

File: docs/formats.rst [hunk 11]
```diff
@@ -7,6 +7,22 @@ Formats
...
+
+The ``cli`` format is currently export-only. The exports produce a representation
+table suited to a terminal.
+
+When exporting to a cli you can pass the table format  with the ``tablefmt``
```
Comment (hugovk) [inline docs/formats.rst]:
```suggestion
When exporting to a CLI you can pass the table format with the ``tablefmt``
```

File: docs/formats.rst [hunk 12]
```diff
@@ -7,6 +7,22 @@ Formats
...
+parameter, the supported formats are the same supported by Pretty-print tabular
+data python-tabulate. For example::
```
Comment (hugovk) [inline docs/formats.rst]: "the supported formats are the same supported by Pretty-print tabular
data python-tabulate"

I don't quite understand this bit, please could you explain or reword?

File: docs/formats.rst [hunk 13]
```diff
@@ -7,6 +7,22 @@ Formats
...
+
+    dataset.export('cli',tablefmt="github")
+    dataset.export('cli',tablefmt="grid")
```
Comment (hugovk) [inline docs/formats.rst]:
```suggestion
    dataset.export("cli", tablefmt="grid")
```

File: docs/formats.rst [hunk 14]
```diff
@@ -7,6 +7,22 @@ Formats
...
+====
+
+The ``cli`` format is currently export-only. The exports produce a representation
+table suited to a terminal.
+
+When exporting to a cli you can pass the table format  with the ``tablefmt``
+parameter, the supported formats are the same supported by Pretty-print tabular
+data python-tabulate. For example::
+
+    dataset.export('cli',tablefmt="github")
```
Comment (hugovk) [inline docs/formats.rst]:
```suggestion
    dataset.export("cli", tablefmt="github")
```

File: src/tablib/formats/__init__.py [hunk 15]
```diff
@@ -124,6 +133,12 @@ def register_builtins(self):
         if find_spec('pandas'):
             self.register('df', 'tablib.formats._df.DataFrameFormat')
         self.register('rst', 'tablib.formats._rst.ReSTFormat')
+<<<<<<< HEAD
+        if find_spec('tabulate'):
+            self.register('cli', 'tablib.formats._cli.CLIFormat')
+=======
+        self.register('cli', 'tablib.formats._cli.CLIFormat')
+>>>>>>> 4fe5cba79fe49e97a417899574fd693e7850795e
```
Comment (hugovk) [inline src/tablib/formats/__init__.py]: Merge conflict needs fixing.

File: src/tablib/formats/_cli.py [hunk 16]
```diff
@@ -0,0 +1,19 @@
+"""Tablib - Command-line Interface table export support.
+
+   Generates a representation for CLI from the dataset.
+   Wrapper for tabulate library
```
Comment (hugovk) [inline src/tablib/formats/_cli.py]:
```suggestion
   Wrapper for tabulate library.
```

**Comment by claudep** [comment]
As soon as @hugovk is satisfied with the PR, so am I.

**Pull request merged**